### PR TITLE
LPD-35168 Binding two trees into one using a relationship

### PR DESCRIPTION
**Pull request by pedro-oliveira446**
[LPD-35168](https://liferay.atlassian.net/browse/LPD-35168)

[LPD-35168]: https://liferay.atlassian.net/browse/LPD-35168?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java [hunk 1]
```diff
@@ -1308,22 +1314,39 @@ private void _bindObjectDefinitions(ObjectRelationship objectRelationship)
...
+
+			ObjectDefinition objectDefinition =
+				objectDefinitionLocalService.fetchObjectDefinition(
+					node.getPrimaryKey());
+
+			String objectDefinitionPreviousRESTContextPath =
+				objectDefinition.getRESTContextPath();
+
+			objectDefinition.setRootObjectDefinitionId(
+				objectDefinition1.getRootObjectDefinitionId());
```
Comment (marcelabc) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java:1333]: From Feliphe's code, there was this condition:
```
if (objectDefinition1.isApproved() == objectDefinition2.isApproved()) {
	objectDefinition2.setRootObjectDefinitionId(
		objectDefinition1.getObjectDefinitionId());
}
else {
	objectDefinition2.setRootObjectDefinitionId(
		objectDefinition2.getObjectDefinitionId());
}
```

Now it is only: 
`objectDefinition.setRootObjectDefinitionId(objectDefinition1.getRootObjectDefinitionId());`

I think it should be:
```
if (objectDefinition1.isApproved() == objectDefinition.isApproved()) {
	objectDefinition.setRootObjectDefinitionId(
		objectDefinition1.getRootObjectDefinitionId());
}
else {
	objectDefinition.setRootObjectDefinitionId(
		objectDefinition.getObjectDefinitionId());
}
```

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java [hunk 2]
```diff
@@ -1308,22 +1314,39 @@ private void _bindObjectDefinitions(ObjectRelationship objectRelationship)
 			_objectDefinitionPersistence.findByPrimaryKey(
 				objectRelationship.getObjectDefinitionId2());
 
-		String objectDefinition2PreviousRESTContextPath =
-			objectDefinition2.getRESTContextPath();
+		Tree tree = _treeFactory.createObjectDefinitionTree(
```
Comment (marcelabc) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java:1317]: Before starting the logic involving objectDefinition2, can we move this code to here:
```
if (objectDefinition1.isApproved()) {
	objectDefinition1.setPreviousRESTContextPath(
		objectDefinition1PreviousRESTContextPath);
	objectDefinitionLocalService.deployObjectDefinition(
		objectDefinition1);
}
```
This method would have a block of logic about objectDefinition1 and after objectDefinition2.
What do you think? can we test it?

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java [hunk 3]
```diff
@@ -514,6 +514,66 @@ public void testBindPublishedObjectDefinitions() throws Exception {
...
+		_createObjectRelationshipEdge(
+			objectDefinitionAAA, objectDefinitionAAAA);
+
+		ObjectDefinition objectDefinitionAAAAA =
+			_addAndPublishCustomObjectDefinition();
+
+		_createObjectRelationshipEdge(
+			objectDefinitionAAAA, objectDefinitionAAAAA);
+
+		_createObjectRelationshipEdge(objectDefinitionAA, objectDefinitionAAA);
```
Comment (marcelabc) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java:542]: Does it make a difference to run
`_createObjectRelationshipEdge(objectDefinitionAAAA, objectDefinitionAAAAA);
`before this
`_createObjectRelationshipEdge(objectDefinitionAA, objectDefinitionAAA); `
to your test scenario?

If not, can we sort and remove the empty line?
```
_createObjectRelationshipEdge(objectDefinitionAA, objectDefinitionAAA);
_createObjectRelationshipEdge(objectDefinitionAAAA, objectDefinitionAAAAA);
```

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java [hunk 4]
```diff
@@ -893,6 +953,28 @@ private ObjectRelationship _addObjectRelationshipSystemObjectDefinition()
 			ObjectRelationshipConstants.TYPE_ONE_TO_MANY, null);
 	}
 
+	private ObjectRelationship _createObjectRelationshipEdge(
```
Comment (marcelabc) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java:956]: Can you show an example of methods in test classes that use the prefix create?

What do you think about `_addObjectRelationshipEdge()` ?

**Review by marcelabc — CHANGES_REQUESTED** [review]
Hi 
I finished my review. Let me know if you have questions :smile:

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java (hunk 2 above)
Comment (marcelabc) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java:1317]: Also, related to this condition, I noticed that @carolmariaabb did this [change](https://github.com/liferay-objects/liferay-portal/pull/3667/commits/a6e9a0b475a636182c39e60a7c2d596d0ed36705) in her PR. Can we please check if it is necessary so you can add it to your PR?

**Comment by pedro-oliveira446** [comment]
ci:test:sf

**Comment by pedro-oliveira446** [comment]
ci:test:sf

**Comment by pedro-oliveira446** [comment]
ci:test:sf

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java [hunk 5]
```diff
@@ -514,6 +517,66 @@ public void testBindPublishedObjectDefinitions() throws Exception {
 					_treeFactory.createObjectDefinitionTree(
 						objectDefinition1.getObjectDefinitionId()),
 					_objectDefinitionLocalService));
+
+		// Bind two published object definitions trees
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java:521]: Bind two published object definitions that belong to a hierarchical structure

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java [hunk 6]
```diff
@@ -514,6 +517,66 @@ public void testBindPublishedObjectDefinitions() throws Exception {
...
+
+		objectDefinitionA = _addAndPublishCustomObjectDefinition();
+		objectDefinitionAA = _addAndPublishCustomObjectDefinition();
+
+		_addObjectRelationshipEdge(objectDefinitionA, objectDefinitionAA);
+
+		ObjectDefinition objectDefinitionAB =
+			_addAndPublishCustomObjectDefinition();
+
+		_addObjectRelationshipEdge(objectDefinitionA, objectDefinitionAB);
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java:531]: Assert that:

````java
TreeTestUtil.assertObjectDefinitionTree(
			LinkedHashMapBuilder.put(
				objectDefinitionA.getShortName(),
				new String[] {
					objectDefinitionAA.getShortName(),
					objectDefinitionAB.getShortName()
				}
			).put(
				objectDefinitionAA.getShortName(),
				new String[] {objectDefinitionAAA.getShortName()}
			).put(
				objectDefinitionAB.getShortName(), new String[0]
			).build(),
			_treeFactory.createObjectDefinitionTree(
				objectDefinitionA.getObjectDefinitionId()),
			_objectDefinitionLocalService);
````

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java [hunk 7]
```diff
@@ -514,6 +517,66 @@ public void testBindPublishedObjectDefinitions() throws Exception {
...
+			_addAndPublishCustomObjectDefinition();
+		ObjectDefinition objectDefinitionAAAA =
+			_addAndPublishCustomObjectDefinition();
+
+		_addObjectRelationshipEdge(objectDefinitionAAA, objectDefinitionAAAA);
+
+		ObjectDefinition objectDefinitionAAAAA =
+			_addAndPublishCustomObjectDefinition();
+
+		_addObjectRelationshipEdge(objectDefinitionAAAA, objectDefinitionAAAAA);
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java:543]: Assert that:

````java
TreeTestUtil.assertObjectDefinitionTree(
			LinkedHashMapBuilder.put(
				objectDefinitionAAA.getShortName(),
				new String[] {
					objectDefinitionAAAA.getShortName()
				}
			).put(
				objectDefinitionAAAA.getShortName(),
				new String[0]
			).build(),
			_treeFactory.createObjectDefinitionTree(
				objectDefinitionAAA.getObjectDefinitionId()),
			_objectDefinitionLocalService);
````

**Comment by pedro-oliveira446** [comment]
ci:test:sf

**Comment by pedro-oliveira446** [comment]
ci:test:relevant

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java (hunk 5 above)
Comment (pedro-oliveira446) [reply modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java]: Done!

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java (hunk 6 above)
Comment (pedro-oliveira446) [reply modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java]: Done!

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java (hunk 7 above)
Comment (pedro-oliveira446) [reply modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java]: Done!

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java [hunk 8]
```diff
@@ -1308,26 +1314,54 @@ private void _bindObjectDefinitions(ObjectRelationship objectRelationship)
...
+
+				nodeObjectDefinition.setPortlet(true);
+			}
+
+			nodeObjectDefinition =
+				objectDefinitionLocalService.updateObjectDefinition(
+					nodeObjectDefinition);
+
+			if (objectDefinition1.isApproved() &&
+				nodeObjectDefinition.isApproved()) {
```
Comment (marcelabc) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java:1359]: Can you sort it please?

**Comment by pedro-oliveira446** [comment]
ci:report:17207599

**Comment by pedro-oliveira446** [comment]
Resent https://github.com/brianchandotcom/liferay-portal/pull/154407

**Pull request closed without merging**